# Assignment for Week 4      
(Chapter 4 & 5)

## Purpose:         
This assignment is intended to help you do the following:
1.	Familiarize with merging datasets together
2.	Learn how to identify missing data from a dataset
3.	Handle the missing data from the dataset


## Action Items:
•	Merge other data to your selected data set.    
1. Find another data set, series, text data or a column from another data source and merge with your selected data set.      
2. Identify the missing values on your data set. Count the missing values on your data set and then clean your data set from the missing values.       
3. Post your work on GitHub     
4. Add a README


In [0]:
# Let's import pandas features and load the datasets
import pandas as pd
import numpy as np
import seaborn as sns
import sys
df = pd.read_csv('https://raw.githubusercontent.com/frankData612/data_612/master/State_Drug_Utilization_Data_2010/State_Drug_Utilization_Data_2010.csv')


In [0]:
# Next, let's laod the second dataset (Stocks_Yahoo)
stocks = pd.read_csv('https://raw.githubusercontent.com/frankdata612/data_612/master/stock_data/stocks_yahoo.csv')

In [5]:
# Let's see the first first rows of the stock dataset
stocks.head(5)

no         date  ... Dividend Yield 1Y target est
0   1  Nov 21 2019  ...      N/A (N/A)      1,487.52
1   2  Nov 21 2019  ...      N/A (N/A)      1,459.33
2   3  Nov 21 2019  ...   6.48 (4.86%)         148.3
3   4  Nov 21 2019  ...   2.04 (1.36%)        160.16
4   5  Nov 21 2019  ...   3.08 (1.16%)        257.21

[5 rows x 19 columns]

In [7]:
# Let's see if there are any null values in the "Stock" dataset
stocks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67905 entries, 0 to 67904
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   no                 67905 non-null  int64  
 1   date               67905 non-null  object 
 2   company_name       67905 non-null  object 
 3   price_at_close     67761 non-null  object 
 4   price_change       67761 non-null  object 
 5   price_after_hours  67761 non-null  object 
 6   previous_close     67761 non-null  object 
 7   today opened       67750 non-null  object 
 8   price range        67756 non-null  object 
 9   52 week range      67657 non-null  object 
 10  volume             67752 non-null  object 
 11  average volume     66584 non-null  object 
 12  market cap         67298 non-null  object 
 13  beta (3Y Monthly)  63163 non-null  float64
 14  PE Ratio           60220 non-null  object 
 15  EPS                64962 non-null  object 
 16  earnings date      576

From the above syntax we can see that several column names have potenital null values (most column names fall short of the 67905 count). 

In [8]:
# Now let's look at the State Drug Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156220 entries, 0 to 156219
Data columns (total 21 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Unnamed: 0                      156220 non-null  int64  
 1   Utilization Type                156220 non-null  object 
 2   State                           156220 non-null  object 
 3   Labeler Code                    156220 non-null  int64  
 4   Product Code                    156220 non-null  int64  
 5   Package Size                    156220 non-null  int64  
 6   Year                            156220 non-null  int64  
 7   Quarter                         156220 non-null  int64  
 8   Product Name                    156166 non-null  object 
 9   Units Reimbursed                156220 non-null  float64
 10  Suppression Used                156220 non-null  bool   
 11  Number of Prescriptions         156220 non-null  int64  
 12  Total Amount Rei

Only the location data from the State Drug Utilization Data seems to have some null values.

When we concatenate DateFrames, we need to specify the axis (axis=0) tells pandas to stack the second DataFrame on the first one. Pandas will automatically detect whether the column names are the same and will stack the data accordingly. Using (axis=1) will stack the columns in the second DataFrame to the "right" of the first DataFrame. To stack the data vertically, we need to make sure we have the same columns and associated column format in both datasets. When we stack horizontally, we want to make sure what we are doing makes sense (that the data points/elements are related in some way ) this ensures value with respect to the merged dataset.

In [12]:
# Let's place the two DataFrames side by side
horizontal_stack = pd.concat([df, stocks], axis=1)
horizontal_stack.info()
horizontal_stack.head(3)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156220 entries, 0 to 156219
Data columns (total 40 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Unnamed: 0                      156220 non-null  int64  
 1   Utilization Type                156220 non-null  object 
 2   State                           156220 non-null  object 
 3   Labeler Code                    156220 non-null  int64  
 4   Product Code                    156220 non-null  int64  
 5   Package Size                    156220 non-null  int64  
 6   Year                            156220 non-null  int64  
 7   Quarter                         156220 non-null  int64  
 8   Product Name                    156166 non-null  object 
 9   Units Reimbursed                156220 non-null  float64
 10  Suppression Used                156220 non-null  bool   
 11  Number of Prescriptions         156220 non-null  int64  
 12  Total Amount Rei

Unnamed: 0 Utilization Type  ... Dividend Yield  1Y target est
0      128037             MCOU  ...      N/A (N/A)       1,487.52
1      491755             FFSU  ...      N/A (N/A)       1,459.33
2     2568076             FFSU  ...   6.48 (4.86%)          148.3

[3 rows x 40 columns]

As you can see from the above syntax we have merged our two datasets (df & stocks) into a new DataFrame called (horizontal_stack). 

In [17]:
horizontal_stack.isnull()

Unnamed: 0  Utilization Type  ...  Dividend Yield  1Y target est
0            False             False  ...           False          False
1            False             False  ...           False          False
2            False             False  ...           False          False
3            False             False  ...           False          False
4            False             False  ...           False          False
...            ...               ...  ...             ...            ...
156215       False             False  ...            True           True
156216       False             False  ...            True           True
156217       False             False  ...            True           True
156218       False             False  ...            True           True
156219       False             False  ...            True           True

[156220 rows x 40 columns]

From our syntax above you can see by the output the we have many missing values denoted by "True" in the column output values.
Next, we will count the missing values.


In [18]:
# Let's count the missing values in our new "horizontal_stack" DataFrame
horizontal_stack.isnull().sum()

Unnamed: 0                            0
Utilization Type                      0
State                                 0
Labeler Code                          0
Product Code                          0
Package Size                          0
Year                                  0
Quarter                               0
Product Name                         54
Units Reimbursed                      0
Suppression Used                      0
Number of Prescriptions               0
Total Amount Reimbursed               0
Medicaid Amount Reimbursed            0
Non Medicaid Amount Reimbursed        0
Quarter Begin                         0
Quarter Begin Date                    0
Latitude                          10804
Longitude                         10804
Location                          10804
NDC                                   0
no                                88315
date                              88315
company_name                      88315
price_at_close                    88459


So, we have many missing values associated with what had been the "stocks" dataset that is now part of our merged " horizontal_stack" new dataset.


Next, we need to handle all the NaN values in our newly combined "horizontal_stack" dataset. Do we do this by replacing the NaN value with "zero", or do we use the "mean" to fill in the gaps in the data values. 

In [0]:
# Let's see what happens if we replace NaN with "zero" using "fillna"
horizontal_stack = horizontal_stack.fillna(0)

In [21]:
horizontal_stack.isnull().sum()

Unnamed: 0                        0
Utilization Type                  0
State                             0
Labeler Code                      0
Product Code                      0
Package Size                      0
Year                              0
Quarter                           0
Product Name                      0
Units Reimbursed                  0
Suppression Used                  0
Number of Prescriptions           0
Total Amount Reimbursed           0
Medicaid Amount Reimbursed        0
Non Medicaid Amount Reimbursed    0
Quarter Begin                     0
Quarter Begin Date                0
Latitude                          0
Longitude                         0
Location                          0
NDC                               0
no                                0
date                              0
company_name                      0
price_at_close                    0
price_change                      0
price_after_hours                 0
previous_close              

So, as you can see from the above syntax all the NaN values in our mereged dataset have been filled with "zero". All counts have been changed to "zero". In theory we can now proceed with anyother data exploration or algorithm use on our combined dataset "horizontal_stack". For example, let's see what a simple statistical analysis gives us.

In [29]:
# Let's see some stats from our newly cleaned "horizontal_stack" dataset
import matplotlib.pyplot as plt
import scipy.stats as stats
import pandas.util.testing as tm
from pandas import set_option

print("SUMMARY STATISTICS OF NUMERIC COLUMNS")
print()
horizontal_stack.describe()


SUMMARY STATISTICS OF NUMERIC COLUMNS



Unnamed: 0  Labeler Code  ...           no  beta (3Y Monthly)
count  1.5622e+05   156220.0000  ...  156220.0000        156220.0000
mean   1.5616e+06    24951.2842  ...   14758.5358             0.4290
std    9.0197e+05    28214.9965  ...   21220.5641             1.5706
min    4.0000e+01        2.0000  ...       0.0000           -90.1100
25%    7.8259e+05      228.0000  ...       0.0000             0.0000
50%    1.5589e+06      781.0000  ...       0.0000             0.0000
75%    2.3427e+06    55111.0000  ...   28850.2500             0.8400
max    3.1244e+06    99207.0000  ...   67905.0000            79.0200

[8 rows x 16 columns]

**Summary:**

Our task this week was to merge two datasets, identify the missing values (NaN), and provide a count of missing values per object in the merged dataset. Then, clean that data preparing it for further exploration/algorithm use. Finally, post the results to our GitHub page.

1 - I merged my State Drug Utilization Data 2010 dataset with the Yahoo_Stocks dataset.

2 - I reviewed/explored those datsets for NaN values, then I did a horizontal stack merging the two datasets.

3 - Next, our new DataFrame "horizontal_stck" was explored for missing values. I used the "isnull().sum()" syntax to show the counts of missing values (NaN) per object with the merged dataset.

4 - I used the "fillna" syntax to replace all NaN values with "zeros" filling in all those gaps in the merged dataset.

5 - Nest, I used the "isnull().sum()" syntax to again show the null value counts per object. This time the output showed no NaN values for any of the objects within our merged DataFrame. 

6 - Afterward, I did a basic output of summary statistics just to do a quick exploration of our newly cleaned DataFrame.